linalg/hypre.cpp

```diff
@@ -2969,22 +2969,37 @@ HypreParMatrix * HypreParMatrixFromBlocks(Array2D<HypreParMatrix*> &blocks,
       }
    }
 
+   MFEM_VERIFY(HYPRE_AssumedPartitionCheck(),
+               "only 'assumed partition' mode is supported");
+
    std::vector<HYPRE_BigInt> rowStarts2(2);
    rowStarts2[0] = first_loc_row;
    rowStarts2[1] = first_loc_row + all_num_loc_rows[rank];
 
-   std::vector<HYPRE_BigInt> colStarts2(2);
-   colStarts2[0] = first_loc_col;
-   colStarts2[1] = first_loc_col + all_num_loc_cols[rank];
+   int square = (first_loc_row == first_loc_col &&
+                 all_num_loc_rows[rank] == all_num_loc_cols[rank]);
+   MPI_Allreduce(MPI_IN_PLACE, &square, 1, MPI_INT, MPI_MAX, comm);
```

Comment:
This looks correct, but I think you can avoid the `MPI_Allreduce` since the sizes are already gathered by `GatherBlockOffsetData` into `all_num_loc_rows` and `all_num_loc_cols`, which are arrays with sizes for all ranks. Also, `first_loc_row` and `first_loc_col` are computed from these arrays, so there is no need to compare them. You can just loop over all ranks `r` and check `all_num_loc_rows[r] == all_num_loc_cols[r]`.

Reply:
Yup, duh. Should be fixed now.